### PR TITLE
Remove the regex from data protection

### PR DESCRIPTION
**Pull request by davidfowl**
Regexes are the devil 😄

### CODE REVIEW

File: src/DataProtection/DataProtection/src/TypeForwardingActivator.cs [hunk 1]
```diff
@@ -64,6 +62,26 @@ internal object CreateInstance(Type expectedBaseType, string originalTypeName, o
...
+
+            var versionEndIndex = forwardedTypeName.IndexOf(',', versionStartIndex);
+
+            if (versionEndIndex == -1)
+            {
+                // No end index?
+                return forwardedTypeName;
+            }
+            
+            return forwardedTypeName.Substring(0, versionStartIndex) + forwardedTypeName.Substring(versionEndIndex + 1);
```
Comment (gfoidl) [inline src/DataProtection/DataProtection/src/TypeForwardingActivator.cs:84]:
```suggestion
            return forwardedTypeName.Substring(0, versionStartIndex) + forwardedTypeName.Substring(versionEndIndex + 2);
```
to eliminate the extra space before `,  Culture=...`

**Comment by gfoidl** [comment]
> 
> 
> Regexes are the devil 😄

Really 🙈 😉?

<details>
  <summary>Simple benchmark</summary>

```c#
using System;
using System.Text.RegularExpressions;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace ConsoleApp1
{
    public static class Program
    {
        public static void Main()
        {
            var bench = new Bench();
            Console.WriteLine(bench.Current());
            Console.WriteLine(bench.PR());
            Console.WriteLine(bench.StringCreate());
#if !DEBUG
            BenchmarkRunner.Run<Bench>();
#endif
        }
    }

    [MemoryDiagnoser]
    public class Bench
    {
        private readonly string _forwardedTypeName = "Type, Assembly, Version=1.2.3.4, Culture={Culture}, PublicKeyToken={PublicKeyToken}";

        [Benchmark(Baseline = true)]
        public string Current() => this.Current(_forwardedTypeName);

        [Benchmark]
        public string PR() => this.PR(_forwardedTypeName);

        [Benchmark]
        public string StringCreate() => this.StringCreate(_forwardedTypeName);

        private static readonly Regex s_versionPattern = new Regex(@",\s?Version=[0-9]+(\.[0-9]+){0,3}", RegexOptions.Compiled, TimeSpan.FromSeconds(2));
        public string Current(string forwardedTypeName) => s_versionPattern.Replace(forwardedTypeName, "");

        public string PR(string forwardedTypeName)
        {
            int versionStartIndex = forwardedTypeName.IndexOf("Version=");

            if (versionStartIndex == -1)
            {
                // No version?
                return forwardedTypeName;
            }

            int versionEndIndex = forwardedTypeName.IndexOf(',', versionStartIndex);

            if (versionEndIndex == -1)
            {
                // No end index?
                return forwardedTypeName;
            }

            return forwardedTypeName.Substring(0, versionStartIndex) + forwardedTypeName.Substring(versionEndIndex + 2);
        }

        public string StringCreate(string forwardedTypeName)
        {
            int versionStartIndex = forwardedTypeName.IndexOf("Version=");

            if (versionStartIndex == -1)
            {
                // No version?
                return forwardedTypeName;
            }

            int versionEndIndex = forwardedTypeName.IndexOf(',', versionStartIndex);

            if (versionEndIndex == -1)
            {
                // No end index?
                return forwardedTypeName;
            }

            int finalLength = versionStartIndex + forwardedTypeName.Length - versionEndIndex - 2;
            return string.Create(finalLength, (forwardedTypeName, versionStartIndex, versionEndIndex), (buffer, state) =>
            {
                ReadOnlySpan<char> source = state.forwardedTypeName.AsSpan();

                source[..state.versionStartIndex].CopyTo(buffer);
                source[(state.versionEndIndex + 2)..].CopyTo(buffer[state.versionStartIndex..]);
            });
        }
    }
}
```
</details>

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.388 (2004/?/20H1)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-preview.7.20366.6
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.36411, CoreFX 5.0.20.36411), X64 RyuJIT
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.36411, CoreFX 5.0.20.36411), X64 RyuJIT


```
|       Method |       Mean |     Error |    StdDev |     Median | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |-----------:|----------:|----------:|-----------:|------:|--------:|-------:|------:|------:|----------:|
|      Current |   572.0 ns |  10.68 ns |   9.47 ns |   571.4 ns |  1.00 |    0.00 | 0.0505 |     - |     - |     160 B |
|           PR | 8,190.1 ns | 163.16 ns | 418.24 ns | 7,970.5 ns | 14.47 |    0.79 | 0.1068 |     - |     - |     344 B |
| StringCreate | 7,916.2 ns | 138.68 ns | 136.20 ns | 7,876.2 ns | 13.85 |    0.37 | 0.0305 |     - |     - |     160 B |

So Regex doesn't look that bad (though I don't know if it comes from recent perf-improvements there or not).
If you'd like to drive out the devil use `string.Create` to avoid the extra allocations.

**Comment by davidfowl** [comment]
This isn't a hot path so I'm not too worried. The dependency on regex is with getting rid of for this tiny piece of code. @stephentoub did too good it a job optimizating regex

**Comment by gfoidl** [comment]
> The dependency on regex is with getting rid

Ah, it's about removing this dependency. Then it's fine 😃

File: src/DataProtection/DataProtection/src/TypeForwardingActivator.cs [hunk 2]
```diff
@@ -64,6 +62,26 @@ internal object CreateInstance(Type expectedBaseType, string originalTypeName, o
         }
 
         protected string RemoveVersionFromAssemblyName(string forwardedTypeName)
-            => _versionPattern.Replace(forwardedTypeName, "");
+        {
+            // Type, Assembly, Version={Version}, Culture={Culture}, PublicKeyToken={Token}
```
Comment (pranavkm) [inline src/DataProtection/DataProtection/src/TypeForwardingActivator.cs:66]: Could we use `AssemblyName` to do this?


```suggestion
            try
            {
               var parsedName = new AssemblyName(typeof(Program).Assembly.FullName) { Version = null };
               return parsedName.ToString();
            }
            catch
            {
               return forwardedTypeName;
            }
```

File: src/DataProtection/DataProtection/src/TypeForwardingActivator.cs (hunk 1 above)
Comment (MaivKhanhPr0) [reply src/DataProtection/DataProtection/src/TypeForwardingActivator.cs:84]: []()

File: src/DataProtection/DataProtection/src/TypeForwardingActivator.cs (hunk 2 above)
Comment (davidfowl) [reply src/DataProtection/DataProtection/src/TypeForwardingActivator.cs:66]: No, you're need to slice off the type name first. I'm considering another approach where I store a list of provider types by type name and assembly as well

**Comment by davidfowl** [comment]
Gonna close this for now.

**Pull request closed without merging**